src/gurt/telemetry.c

```diff
@@ -3155,14 +3140,91 @@ d_tm_list(struct d_tm_context *ctx, struct d_tm_nodeList_t **head,
 	}
 
 	while (node != NULL) {
-		rc = d_tm_list(ctx, head, node, d_tm_type);
+		rc = list_children(ctx, head, node, d_tm_type,
+				   cur_depth, max_depth, 0);
 		if (rc != DER_SUCCESS)
 			goto out;
 		node = conv_ptr(shmem, node->dtn_sibling);
 	}
 
 out:
 	return rc;
+
+}
+
+/**
+ * Perform a recursive listing from the given \a node for all subdirectories
+ * to a maximum depth of \a max_depth.
+ *
+ * \param[in]		ctx		Telemetry context
+ * \param[in,out]	head		Pointer to a nodelist
+ * \param[in]		node		The recursive listing starts
+ *					from this node.
+ * \param[in,out]	num_dirs	Optional pointer to storage for the
+ *					number of directories found.
+ * \param[in]		max_depth	The maximum depth of the listing, including
+ *					the root node.
+ *
+ * \return		DER_SUCCESS		Success
+ *			-DER_NOMEM		Out of global heap
+ *			-DER_INVAL		Invalid pointer for \a head or
+ *						\a node
+ */
+int
+d_tm_list_subdirs(struct d_tm_context *ctx, struct d_tm_nodeList_t **head,
+		  struct d_tm_node_t *node, uint64_t *num_dirs, int max_depth)
+{
+	int			 rc = DER_SUCCESS;
+	uint64_t		 dir_count = 0;
+	struct d_tm_nodeList_t	*cur_head = NULL;
+
+	rc = list_children(ctx, head, node, D_TM_DIRECTORY, 0, max_depth, 1);
+	if (rc != DER_SUCCESS)
+		return rc;
+
+	cur_head = *head;
+	while (cur_head != NULL && cur_head->dtnl_node != NULL) {
+		if (cur_head->dtnl_node->dtn_type == D_TM_DIRECTORY)
+			dir_count++;
+
+		if (cur_head->dtnl_next == NULL)
+			break;
```

Comment:
Without this it should be caught by the while loop on the next iteration

Reply:
Ah, yep. I added the check on cur_head in the while loop after I added this.